### PR TITLE
redesign of mime crate

### PR DESCRIPTION
**Pull request by seanmonstar**
The redesign has these goals in mind:

- **It should be fast.** The improvement here is that the full text is stored in a single buffer. In the previous versions, being a bunch of internal enums, it was quite slow to use `fmt::Display`. Writing a single buffer reduces the overhead of the `std::fmt` machinery.
  
  Comparisons with constants is near instant. Comparing with parsed `Mime`s is slightly slower, but there is also room to add optimizations which I've skipped in order to get this our the door.
- **It should be easy** to create a `Mime`. In this design, you can use one of the common constants, like `mime::TEXT_PLAIN`, or parse a custom string into a `Mime`, like `"application/vnd.github.api.v3+json".parse()`. No more importing 5 different types or passing empty `Vec`s.
 
  With macros 2.0, it will be possible to make a macro that can parse the text at compile time, allowing custom constant `Mime`s.
- **It should be extensible.** As a bunch of enums, adding new variants was technically a breaking change (though who would match on every `SubLevel` variant?). By using constants, new ones can be added without breaking anyone.
- **Knowledge of suffixes.** You can check the suffix, such as in `image/svg+xml`.
- **Comparing is still easy.** You can still compare the various components of a `Mime` in a match. This works perfectly fine:
  
  ```rust
  match (mime.type_(), mime.subtype()) {
      (mime::TEXT, mime::PLAIN) => println!("plain text"),
      (mime::TEXT, _) => println!("structured text"),
      _ => println!("not text"),
  }
  ```
  
  Additionally, `Mime` and its components can be easily compared against `&str`.

  ```rust
  assert_eq!(mime::TEXT_PLAIN, "text/plain");
  ```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -5,651 +5,741 @@
...
+    Custom(usize, Vec<(Str, Str)>),
+    None,
+}
 
-        #[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd)]
-        pub enum $en {
-            $($ty),*,
-            $ext(String)
+#[derive(Clone, Copy)]
+struct Str(usize, usize);
```
Comment (robinst) [inline src/lib.rs]: The naming of this threw me off a bit until I saw the definition. I wonder if `Slice` or something else that's a bit more different from `str` than `Str` would be better?

File: src/lib.rs [hunk 2]
```diff
@@ -5,651 +5,741 @@
...
+    TEXT_JAVSCRIPT, "text/javascript", 4;
+    TEXT_XML, "text/xml", 4;
+    TEXT_EVENT_STREAM, "text/event-stream", 4;
+
+    IMAGE_JPEG, "image/jpeg", 5;
+    IMAGE_GIF, "image/gif", 5;
+    IMAGE_PNG, "image/png", 5;
+    IMAGE_BMP, "image/bmp", 5;
+
+    APPLICATION_JSON, "application/json", 11;
```
Comment (robinst) [inline src/lib.rs]: `application/json; charset=utf-8` is also common, would be nice to include as well.

File: src/parse.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+];
+
+fn is_restricted_name_first_char(c: u8) -> bool {
+    RESTRICTED_NAME_FIRST[c as usize]
```
Comment (robinst) [inline src/parse.rs]: (Is this faster than the corresponding `match` statement? I somehow expected Rust/LLVM to understand that it's matching on a u8 and generate this table itself.)

Comment (seanmonstar) [inline src/parse.rs]: LLVM does not generate a table, but many instead many comparisons. Running the parse benchmarks with the table and without showed (on my machine) ~20% improvement using the table.

File: src/lib.rs (hunk 2 above)
Comment (seanmonstar) [inline src/lib.rs]: And it's wrong 😂 http://www.iana.org/assignments/media-types/application/json

File: src/lib.rs (hunk 1 above)
Comment (seanmonstar) [inline src/lib.rs]: I agree, the name here is not good. I think I'll switch to `Indices` or `Indexed`.

**Comment by seanmonstar** [comment]
@robinst thanks for looking! That's why I put the link in a bunch of places, since it's such a different design, I wanted a sort of warning to all, and if anyone saw a glaring hole in the new API.

File: src/lib.rs (hunk 2 above)
Comment (robinst) [inline src/lib.rs]: Oh, thanks for the link, that's news :). I definitely remember having to add the charset to make some implementations work, but I guess the implementation was technically wrong in that case.

So it's probably better to not add it to guide people to the correct one. When someone really needs it they can just write it out themselves.

**Pull request merged**